layer/executables/packager.py

```diff
@@ -0,0 +1,154 @@
+
+        target = (output_dir or Path(".")) / executable_name
+
+        # create the executable
+        zipapp.create_archive(
+            source, target, interpreter="/usr/bin/env python", compressed=True
+        )
+
+        # ensure the archive is executable
+        target.chmod(0o744)
+
+        return target
+
+
+def _package_resources(source: Path, resources: List[Path]) -> None:
+    resource_root = source / "resources"
+    resource_root.mkdir(parents=True, exist_ok=True)
+
+    def _copy_resource_file(resource_file: Path) -> None:
+        target_path = resource_root / resource_file
+        target_path.parent.mkdir(parents=True, exist_ok=True)
+        shutil.copyfile(resource_file, target_path)
+
+    def _walk_dir(dir: Path) -> None:
+        for root, dirs, files in os.walk(dir):
+            root_path = Path(root)
+            for file in files:
+                _copy_resource_file(root_path / file)
+            for dir_ in dirs:
+                _walk_dir(root_path / dir_)
+
+    for resource in resources:
+        if resource.is_file():
+            _copy_resource_file(resource)
+        elif resource.is_dir():
+            _walk_dir(resource)
+
+
+# depenedencies required by the runtime itself
+_runtime_requirements = ["cloudpickle==2.1.0"]
+
+
+def _loader_source() -> str:
+    """Returns the source code of the layer function loader."""
+
+    # entrypoint function body is inlined in the executable's __main__.py
+    def _entrypoint() -> None:
+        # pylint: disable=W0404
+        import os
+        import shutil
+        import subprocess  # nosec
+        import tempfile
+        import venv
+        import zipfile
+        from pathlib import Path
+
+        with tempfile.TemporaryDirectory() as env_dir:
+            # create virtual environment
+            venv.create(env_dir, with_pip=True)
+
+            python_path = os.path.join(env_dir, "bin", "python")
+
+            with zipfile.ZipFile(os.path.dirname(__file__)) as exec:
+                # install dependencies
+                with exec.open("requirements.txt", mode="r") as requirements:
+                    pip_install = [
+                        python_path,
+                        "-m",
+                        "pip",
+                        "--quiet",
+                        "--disable-pip-version-check",
+                        "install",
+                    ] + [package.decode("utf-8") for package in requirements]
+                subprocess.check_call(pip_install)  # nosec
+
+                # extract resources
+                for entry in exec.infolist():
+                    if entry.filename.startswith("resources/"):
+                        with exec.open(entry, mode="r") as resource:
+                            target_path = (
+                                Path(".") / entry.filename[len("resources/") :]
+                            )
+                            target_path.parent.mkdir(parents=True, exist_ok=True)
```

Comment:
should this be moved into the `if` block below?

Reply:
No, dirs are always created even though they're empty.